## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd
import numpy

df = pd.read_csv('Resources/client_dataset.csv')
df.head(100)

first      last                                  job  \
0      Donald   Harding                         Immunologist   
1     Tiffany     Myers                      Music therapist   
2     Shannon    Watson                         Immunologist   
3      Nathan     Baker                Accounting technician   
4   Christina  Schwartz                         Chiropractor   
..        ...       ...                                  ...   
95       Erik     Gross                           IT trainer   
96       Kyle  Williams  Engineer, broadcasting (operations)   
97       Nina    Mullen                        Warden/ranger   
98   Nicholas  Williams                       Radio producer   
99  Katherine    Hughes      Commercial/residential surveyor   

                    phone                                     email  \
0      793-904-7725x39308          harding.donald.7185@sullivan.com   
1        201.442.4543x942         myers.t.6537@ferguson-johnson.net   
2       687.737.9424x8503                     swatson8146@payne.net   
3        827-788-8123x012                    bakernathan@benson.com   
4            265-829-3643       christinaschwartz9252@mcconnell.com   
..                    ...                                       ...   
95      313-970-3023x5865                      erikgross@nelson.com   
96        +1-432-234-2942  communitywilliams9024@taylor-padilla.biz   
97  001-595-635-1880x1626                   mullennina3963@fox.info   
98      393.560.7337x2039         nicholas.williams.6317@strong.biz   
99      984.810.5078x1155                     hughesk9561@brown.org   

    client_id  order_id  order_date  order_week  order_year         item_id  \
0       58515   8953482  2023-04-28          17        2023  EUD29711-63-6U   
1       37609   8069089  2023-05-19          20        2023  XDA18116-89-4A   
2       57113   1902144  2023-01-29           4        2023  ABE59463-05-7E   
3       46554   9031802  2023-04-25          17        2023  ZMM00836-65-0C   
4       92089   1322274  2023-05-28          21        2023  BZX55559-12-3X   
..        ...       ...         ...         ...         ...             ...   
95      32098   3998530  2023-05-20          20        2023  EUD29711-63-6U   
96      81012   8126485  2023-05-17          20        2023  AUM78855-69-8C   
97      71958   2164235  2023-03-08          10        2023  DDM14031-55-4U   
98      56590   7777768  2023-01-15           2        2023  UZX16759-11-7E   
99      84527   7272566  2023-02-01           5        2023  ZDX29223-32-0B   

       category         subcategory  unit_price  unit_cost  unit_weight   qty  \
0         decor            wall art     1096.80     762.71         7.50   105   
1   consumables                pens       24.95      15.09         1.49    21   
2      software  project management       13.52       7.86         1.68    39   
3   consumables                pens       36.42      24.85         1.23    29   
4   consumables                misc      195.10     108.17        46.43    20   
..          ...                 ...         ...        ...          ...   ...   
95        decor            wall art     1096.80     762.71         7.50   107   
96  consumables   bathroom supplies       63.78      37.88         3.80    45   
97    furniture             seating       40.22      22.70         1.04    32   
98     software          accounting       13.68      10.71         7.95  1167   
99  consumables      paper products      199.05     183.99         8.11    22   

    line_number  
0             1  
1             0  
2             6  
3             3  
4             1  
..          ...  
95            3  
96            2  
97            6  
98            4  
99            4  

[100 rows x 18 columns]

In [2]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe(include=numpy.number)

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
print("Unique clients: " + str(len(df["client_id"].unique())))
print("Unique items: " + str(len(df["item_id"].unique())))
print("Jobs: " + df["job"].unique())

Unique clients: 999
Unique items: 500
['Jobs: Immunologist' 'Jobs: Music therapist'
 'Jobs: Accounting technician' 'Jobs: Chiropractor'
 'Jobs: Production designer, theatre/television/film' 'Jobs: Barista'
 'Jobs: Theme park manager' 'Jobs: Chiropodist' 'Jobs: Secretary, company'
 'Jobs: Engineer, broadcasting (operations)'
 'Jobs: Surveyor, land/geomatics' 'Jobs: Lobbyist'
 'Jobs: Psychologist, clinical' 'Jobs: Research scientist (maths)'
 'Jobs: Government social research officer' 'Jobs: Colour technologist'
 'Jobs: Production assistant, television'
 'Jobs: Lecturer, further education'
 'Jobs: Commercial art gallery manager' 'Jobs: Architectural technologist'
 'Jobs: Engineer, manufacturing systems' 'Jobs: Textile designer'
 'Jobs: Air cabin crew' 'Jobs: Medical illustrator'
 'Jobs: Technical brewer' 'Jobs: Radiographer, therapeutic'
 'Jobs: Race relations officer' 'Jobs: Horticulturist, amenity'
 'Jobs: Surveyor, rural practice' 'Jobs: Science writer'
 'Jobs: Airline pilot' 'Jobs: E

In [5]:
# What three item categories had the most entries?
df.value_counts("category").head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entries?
df[df["category"] == df.value_counts("category").index[0]].value_counts("subcategory").head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
df.value_counts("client_id").head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
topfive = df.value_counts("client_id").head(5).index.tolist()
topfive

[33615, 66037, 46820, 38378, 24741]

In [9]:
# How many total units (the qty column) did the client with the most entries order order?
print(f"The client with client ID {topfive[0]} ordered " + str(df[df["client_id"] == topfive[0]]["qty"].sum()) + " units")

The client with client ID 33615 ordered 64313 units


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["subtotal"] = df["unit_price"]*df["qty"]

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

df["shipping_price"] = (df["unit_weight"]*df["qty"]).apply(lambda x: x*7 if x>50 else x*10)

df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

df["total_price"] = ((df["subtotal"] + df["shipping_price"]) * (1 + 9.25 / 100)).round(2)

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping_price  \
0     762.71         7.50  105            1  115164.00         5512.50   
1      15.09         1.49   21            0     523.95          312.90   
2       7.86         1.68   39            6     527.28          458.64   
3      24.85         1.23   29            3    1056.18          356.70   
4     108.17        46.43   20            1    3902.00         6500.20   

   total_price  
0    131839.08  
1       914.26  
2      1077.12  
3      1543.57  
4     11364.40  

[5 rows x 21 columns]

In [13]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df["cost"] = df["unit_cost"] * df["qty"] + df["shipping_price"]

df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price  total_price  \
0         7.50  105            1  115164.00         5512.50    131839.08   
1         1.49   21            0     523.95          312.90       914.26   
2         1.68   39            6     527.28          458.64      1077.12   
3         1.23   29            3    1056.18          356.70      1543.57   
4        46.43   20            1    3902.00         6500.20     11364.40   

       cost  
0  85597.05  
1    629.79  
2    765.18  
3   1077.35  
4   8663.60  

[5 rows x 22 columns]

In [14]:
# Create a column for the profit of each line using line cost and line price

df["profit"] = df["total_price"] - df["cost"]

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_price  total_price      cost    profit  
0            1  115164.00         5512.50    131839.08  85597.05  46242.03  
1            0     523.95          312.90       914.26    629.79    284.47  
2            6     527.28          458.64      1077.12    765.18    311.94  
3            3    1056.18          356.70      1543.57   1077.35    466.22  
4            1    3902.00         6500.20     11364.40   8663.60   2700.80  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
print("Order ID 2742071 had a total price of $" + str(df[df["order_id"] == 2742071]["total_price"].sum().round(2)))
print("Order ID 2173913 had a total price of $" + str(df[df["order_id"] == 2173913]["total_price"].sum().round(2)))
print("Order ID 6128929 had a total price of $" + str(df[df["order_id"] == 6128929]["total_price"].sum().round(2)))

Order ID 2742071 had a total price of $152811.89
Order ID 2173913 had a total price of $162388.71
Order ID 6128929 had a total price of $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
[print(f"Client {topfive[k]} spent $" + str(df[df["client_id"]==topfive[k]]["total_price"].sum().round(2))) for k in range(0,5)]

Client 33615 spent $8377308.52
Client 66037 spent $10259514.79
Client 46820 spent $9743794.36
Client 38378 spent $12906550.87
Client 24741 spent $82268892.02


[None, None, None, None, None]

In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

def m(x):
    return {"cid":x,
            "tu":df[df["client_id"]==x]["qty"].sum(),
            "ts":df[df["client_id"]==x]["shipping_price"].sum().round(2),
            "tr":df[df["client_id"]==x]["total_price"].sum().round(2),
            "tp":df[df["client_id"]==x]["profit"].sum().round(2)
    }

d = []

[d.append(m(k)) for k in topfive]

d = pd.DataFrame(d)

d

cid      tu          ts           tr           tp
0  33615   64313  1828984.89   8377308.52   2201994.61
1  66037   43018  1395151.85  10259514.79   3255031.81
2  46820   75768  1601448.84   9743794.36   2736602.72
3  38378   73667  3429455.40  12906550.87   3271829.89
4  24741  239862  5126448.37  82268892.02  36579992.31

In [18]:
# Format the data and rename the columns to names suitable for presentation.
p = d.copy()
p.columns = ["Client ID","Total Units","Total Shipping","Total Revenue","Total Profit"]

# Define the money columns. 
p.rename(columns={"Total Shipping":"Total Shipping ($)", "Total Revenue":"Total Revenue ($)", "Total Profit":"Total Profit ($)"}, inplace=True)

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return x/(10**6)

# Apply the currency_format_millions function to only the money columns. 
p["Total Shipping ($)"] = p["Total Shipping ($)"].apply(currency_format_millions).round(2)
p["Total Revenue ($)"] = p["Total Revenue ($)"].apply(currency_format_millions).round(2)
p["Total Profit ($)"] = p["Total Profit ($)"].apply(currency_format_millions).round(2)

# Rename the columns to reflect the change in the money format. 
p.rename(columns={"Total Shipping ($)":"Total Shipping ($M)", "Total Revenue ($)":"Total Revenue ($M)", "Total Profit ($)":"Total Profit ($M)"}, inplace=True)

p



Client ID  Total Units  Total Shipping ($M)  Total Revenue ($M)  \
0      33615        64313                 1.83                8.38   
1      66037        43018                 1.40               10.26   
2      46820        75768                 1.60                9.74   
3      38378        73667                 3.43               12.91   
4      24741       239862                 5.13               82.27   

   Total Profit ($M)  
0               2.20  
1               3.26  
2               2.74  
3               3.27  
4              36.58

In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
ps = p.sort_values(by="Total Profit ($M)",ascending=False)
ps

Client ID  Total Units  Total Shipping ($M)  Total Revenue ($M)  \
4      24741       239862                 5.13               82.27   
3      38378        73667                 3.43               12.91   
1      66037        43018                 1.40               10.26   
2      46820        75768                 1.60                9.74   
0      33615        64313                 1.83                8.38   

   Total Profit ($M)  
4              36.58  
3               3.27  
1               3.26  
2               2.74  
0               2.20

In [20]:
"""
The findings are that client 24741 is the biggest consumer by revenue of our products with over ten times the profit of any other client.
Even though client 33615 had the most entries in the dataset, client 24741 has the most revenue.
"""

'\nThe findings are that client 24741 is the biggest consumer by revenue of our products with over ten times the profit of any other client.\nEven though client 33615 had the most entries in the dataset, client 24741 has the most revenue.\n'